proxy/src/main/java/com/wavefront/agent/PostPushDataTimedTask.java

```diff
@@ -122,30 +138,60 @@ public UUID getDaemonId() {
   }
 
   public PostPushDataTimedTask(String pushFormat, ForceQueueEnabledAgentAPI agentAPI, String logLevel,
-                               UUID daemonId, String handle, int threadId) {
+                               UUID daemonId, String handle, int threadId, RecyclableRateLimiter pushRateLimiter,
+                               long pushFlushInterval) {
     this.pushFormat = pushFormat;
     this.logLevel = logLevel;
     this.daemonId = daemonId;
     this.handle = handle;
+    this.threadId = threadId;
+    this.pushFlushInterval = pushFlushInterval;
     this.agentAPI = agentAPI;
+    this.pushRateLimiter = pushRateLimiter;
 
     this.pointsAttempted = Metrics.newCounter(new MetricName("points." + handle, "", "sent"));
     this.pointsQueued = Metrics.newCounter(new MetricName("points." + handle, "", "queued"));
     this.pointsBlocked = Metrics.newCounter(new MetricName("points." + handle, "", "blocked"));
     this.pointsReceived = Metrics.newCounter(new MetricName("points." + handle, "", "received"));
-    this.batchesSent = Metrics.newCounter(
+    this.permitsGranted = Metrics.newCounter(new MetricName("limiter", "", "permits-granted"));
+    this.permitsDenied = Metrics.newCounter(new MetricName("limiter", "", "permits-denied"));
+    this.permitsRetried = Metrics.newCounter(new MetricName("limiter", "", "permits-retried"));
+    this.batchesAttempted = Metrics.newCounter(
         new MetricName("push." + String.valueOf(handle) + ".thread-" + String.valueOf(threadId), "", "batches"));
     this.batchSendTime = Metrics.newTimer(new MetricName("push." + handle, "", "duration"),
         TimeUnit.MILLISECONDS, TimeUnit.MINUTES);
+    this.scheduler.schedule(this, pushFlushInterval, TimeUnit.MILLISECONDS);
   }
 
   @Override
   public void run() {
+    long nextRunMillis = this.pushFlushInterval;
     try {
-      List<String> current = createAgentPostBatch();
-      batchesSent.inc();
+      int batchSize;
+      List<String> current = null;
+      batchesAttempted.inc();
+      numIntervals++;
+
+      synchronized (pointsMutex) {
+        batchSize = Math.min(points.size(), pointsPerBatch);
+        if (batchSize != 0) {
+          if (!pushRateLimiter.tryAcquire(batchSize)) {
```

Comment:
acquiring a permit while holding the pointsMutex lock seems fishy. can you comment in your reasoning here?